src/Helper/PaymentHelper.php

```diff
@@ -34,19 +53,126 @@ class PaymentHelper {
+    }
+    /*
+     * The paymentReferenceField is not a part of the form submission,
+     * so we get it from the POST payload.
+     */
+    $request = $this->requestStack->getCurrentRequest();
+    $paymentReferenceField = $request->request->get('os2forms_payment_reference_field');
+
+    /*
+     * The goal here is to store the payment_id, amount_to_pay and status
+     * as a JSON object in the os2forms_payment submission value.
+     */
+
+    $amountToPay = $this->getAmountToPay($submissionData, $paymentElement['#amount_to_pay']);
+
+    if ($request && $amountToPay) {
+
+      $this->updateWebformSubmissionPaymentObject($submission, NULL, NULL, [
+        'payment_id' => $paymentReferenceField,
+        'amount' => $amountToPay,
+        'status' => 'not charged',
+      ]);
+
+    }
+  }
+
+  /**
+   * Updates the payment object in a webform submission.
+   *
+   * @param \Drupal\webform\WebformSubmissionInterface $webformSubmission
+   *   The webform submission to update.
+   * @param string|null $key
+   *   The key of the payment object to update.
+   * @param mixed|null $value
+   *   The value to set for the payment object key.
+   * @param array|null $paymentObject
+   *   The payment object to replace the existing with.
+   *
+   * @throws \Drupal\Core\Entity\EntityStorageException
+   *   Throws Exception.
+   */
+  public function updateWebformSubmissionPaymentObject(WebformSubmissionInterface $webformSubmission, ?string $key = NULL, mixed $value = NULL, ?array $paymentObject = NULL): WebformSubmissionInterface {
+    $submissionData = $webformSubmission->getData();
+    $paymentElementMachineName = $this->findPaymentElement($webformSubmission);
+
+    if ($paymentElementMachineName !== NULL) {
+      if ($paymentObject !== NULL) {
+        $submissionData[$paymentElementMachineName] = json_encode(['paymentObject' => $paymentObject]);
+      }
+      elseif ($key !== NULL && $value !== NULL) {
+        $paymentData = json_decode($submissionData[$paymentElementMachineName], TRUE);
+        $paymentData['paymentObject'][$key] = $value;
+        $submissionData[$paymentElementMachineName] = json_encode($paymentData);
+      }
+      $webformSubmission->setData($submissionData);
+    }
+    return $webformSubmission;
+  }
+
+  /**
+   * Finds the payment element within a webform submission.
+   *
+   * @param \Drupal\webform\WebformSubmissionInterface $submission
+   *   The webform submission object.
+   *
+   * @return string|null
+   *   The key of the payment element if found, or NULL if not found.
+   */
+  private function findPaymentElement(WebformSubmissionInterface $submission) {
     $webformElements = $submission->getWebform()->getElementsDecodedAndFlattened();
+
+    foreach ($webformElements as $elementKey => $webformElement) {
+      if ('os2forms_payment' === ($webformElement['#type'] ?? NULL)) {
+        return $elementKey;
+      }
+    }
+
+  }
```

Comment:
Rename to reflect what the function actually does, add return type and explicit return at end of function:

```suggestion
  private function findPaymentElementKey(WebformSubmissionInterface $submission): ?string {
    $webformElements = $submission->getWebform()->getElementsDecodedAndFlattened();

    foreach ($webformElements as $elementKey => $webformElement) {
      if ('os2forms_payment' === ($webformElement['#type'] ?? NULL)) {
        return $elementKey;
      }
    }

    return NULL;
  }
```